module/zfs/vdev_raidz.c

```diff
@@ -2744,6 +2842,175 @@ vdev_raidz_worst_error(raidz_row_t *rr)
+#define	LAT_OUTLIER_LIMIT	50
+
+static int
+latency_compare(const void *arg1, const void *arg2)
+{
+	const uint64_t *l1 = (uint64_t *)arg1;
+	const uint64_t *l2 = (uint64_t *)arg2;
+
+	return (TREE_CMP(*l1, *l2));
+}
+
+/*
+ * Check for any latency outlier from latest set of child reads.
+ *
+ * Uses a Tukey's fence, with K = 2, for detecting extreme outliers. This
+ * rule defines extreme outliers as data points outside the fence of the
+ * third quartile plus two times the Interquartile Range (IQR). This range
+ * is the distance between the first and third quartile.
+ */
+static void
+vdev_child_slow_outlier(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	if (raidz_read_sit_out_secs == 0 || vd->vdev_children < LAT_SAMPLES_MIN)
+		return;
+
+	spa_t *spa = zio->io_spa;
+	if (spa_load_state(spa) == SPA_LOAD_TRYIMPORT ||
+	    spa_load_state(spa) == SPA_LOAD_RECOVER ||
+	    (spa_load_state(spa) != SPA_LOAD_NONE &&
+	    spa->spa_last_open_failed)) {
+		return;
+	}
+
+	hrtime_t now = gethrtime();
+	uint64_t last = atomic_load_64(&vd->vdev_last_latency_check);
+
+	if ((now - last) < MSEC2NSEC(raid_outlier_check_interval_ms) ||
+	    atomic_cas_64(&vd->vdev_last_latency_check, last, now) != last) {
+		return;
+	}
+
+	int samples = vd->vdev_children;
+	uint64_t data[LAT_SAMPLES_STACK];
+	uint64_t *lat_data;
+
+	if (samples > LAT_SAMPLES_STACK)
+		lat_data = kmem_alloc(sizeof (uint64_t) * samples, KM_SLEEP);
+	else
+		lat_data = &data[0];
+
+	uint64_t max = 0;
+	uint64_t max_outier_count = 0;
+	vdev_t *svd = NULL; /* suspect vdev */
+	vdev_t *ovd = NULL; /* largest outlier vdev */
+	for (int c = 0; c < samples; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+
+		if (cvd->vdev_read_sit_out_expire != 0) {
+			if (cvd->vdev_read_sit_out_expire < gethrtime()) {
+				/*
+				 * Done with our sit out, wait for new outlier
+				 * to emerge.
+				 */
+				cvd->vdev_read_sit_out_expire = 0;
+			} else {
+				atomic_store_64(&cvd->vdev_ewma_latency, 0);
+				/* Only one sit out disk at a time for now */
+				goto out;
+			}
+		}
+
+		lat_data[c] = atomic_load_64(&cvd->vdev_ewma_latency);
+
+		/* wait until all disks have been read from */
+		if (lat_data[c] == 0)
+			goto out;
+
+		/* keep track of the vdev with largest value */
+		if (lat_data[c] > max) {
+			max = lat_data[c];
+			svd = cvd;
+		}
+
+		uint64_t count = atomic_load_64(&cvd->vdev_outlier_count);
+		if (count > max_outier_count) {
+			max_outier_count = count;
+			ovd = cvd;
+		}
+	}
+
+	qsort((void *)lat_data, samples, sizeof (uint64_t), latency_compare);
+	uint64_t fence = latency_quartiles_fence(lat_data, samples);
+	if (lat_data[samples - 1] > fence) {
+		/*
+		 * Keep track of how many times this child has had
+		 * an outlier read. A disk that persitently has a
+		 * higher than peers outlier count will be considered
+		 * a slow disk.
+		 */
+		if (atomic_add_64_nv(&svd->vdev_outlier_count, 1) >
+		    LAT_OUTLIER_LIMIT && svd == ovd &&
+		    svd->vdev_read_sit_out_expire == 0) {
```

Comment:
I am not sure why do you care about the vdev with maximum number of outliers here, considering first of them reachinf `LAT_OUTLIER_LIMIT` will reset all to 0?  Also since this code should be executed only by one thread, I am not sure after `atomic_cas_64()` above `vdev_outlier_count` needs atomic accesses. `vdev_read_sit_out_expire` should also be always 0 here, since you've checked that above.